logutils/logutils.go

```diff
@@ -255,42 +306,243 @@ func shouldSkipFrame(frame runtime.Frame) bool {
 		strings.LastIndex(frame.File, "entry.go") > 0
 }
 
-// StreamHook is a logrus Hook that writes to a stream when fired.
-// It supports configuration of log levels at which is fires.
-type StreamHook struct {
-	mu     sync.Mutex
-	writer io.Writer
-	levels []log.Level
+type QueuedLog struct {
+	Level         log.Level
+	Message       []byte
+	SyslogMessage string
+	WaitGroup     *sync.WaitGroup
+}
+
+func (ql QueuedLog) OnLogDone() {
+	if ql.WaitGroup != nil {
+		ql.WaitGroup.Done()
+	}
+}
+
+func NewStreamDestination(level log.Level, writer io.Writer) *StreamDestination {
+	return &StreamDestination{
+		level:   level,
+		writer:  writer,
+		channel: make(chan QueuedLog, 10000),
+	}
+}
+
+type StreamDestination struct {
+	level   log.Level
+	writer  io.Writer
+	channel chan QueuedLog
+
+	// Our own copy of the dropped logs counter, used for logging out when we drop logs.
+	// Must be read/updated using atomic.XXX.
+	numDroppedLogs uint64
 }
 
-func (h *StreamHook) Levels() []log.Level {
+func (d *StreamDestination) Level() log.Level {
+	return d.level
+}
+
+func (d *StreamDestination) Channel() chan<- QueuedLog {
+	return d.channel
+}
+
+func (d *StreamDestination) OnLogDropped() {
+	atomic.AddUint64(&d.numDroppedLogs, 1)
+}
+
+func (d *StreamDestination) LoopWritingLogs() {
+	var numSeenDroppedLogs uint64
+	for ql := range d.channel {
+		currentNumDroppedLogs := atomic.LoadUint64(&d.numDroppedLogs)
+		if currentNumDroppedLogs > numSeenDroppedLogs {
+			fmt.Fprintf(d.writer, "... dropped %d logs ...\n",
+				currentNumDroppedLogs-numSeenDroppedLogs)
```

Comment:
Yes, I think that's a fair point that the log could be confusingly out-of-order.  I'll take another pass at it.